lib/client.ml

```diff
@@ -28,10 +28,18 @@ let pp_event ppf = function
   | `Channel_exit_status (id, r) -> Format.fprintf ppf "exit %lu with %lu" id r
   | `Disconnected -> Format.fprintf ppf "disconnected"
 
+type kex_state =
+  | Negotiated_kex of string * Ssh.kexinit * string * Ssh.kexinit * Kex.negotiation * Mirage_crypto_pk.Dh.secret * Ssh.mpint
+
+type gex_state =
+  | Requested_gex of string * Ssh.kexinit * string * Ssh.kexinit * Kex.negotiation * int32 * int32 * int32
+  | Negotiated_gex of string * Ssh.kexinit * string * Ssh.kexinit * Kex.negotiation * int32 * int32 * int32 * Z.t * Z.t * Mirage_crypto_pk.Dh.secret * Ssh.mpint
```

Comment:
Is it worth putting an inline record in here to distinguish `cv, ckex, sv, skex, neg, sec, mypub`? it would simplify passing the function args through to handle_kexdh_reply as well

Reply:
Thanks for your suggestion. I'm trying it out:
```OCaml
type kex_state =
  | Negotiated_kex of { client_version : string ; client_kex : Ssh.kexinit ;
                        server_version : string ; server_kex : Ssh.kexinit ;
                        negotiated : Kex.negotiation ;
                        dh_secret : Mirage_crypto_pk.Dh.secret ;
                        dh_share : Ssh.mpint }
```

and then in the pattern match (`input_msg`):
```OCaml
  | Kex (Negotiated_kex data),
    Msg_kex (i, d) ->
    begin
      Wire.dh_kexdh_of_kex i d >>= function
      | Msg_kexdh_reply (pub, theirs, signed) ->
        handle_kexdh_reply t now data pub theirs signed
      | _ ->
        Error "unexpected KEX message"
    end
```

and `handle_kexdh_reply` receives a record. I get the following error (the error location is the `data` passed to `handle_kexdh_reply`):
```
This form is not allowed as the type of the inlined record could escape.
```

So I'd need to destruct the inline record directly in the pattern match, and pass the fields one-by-one to the `handle_kexdh_reply`? Or is there another way to pass the record?
IMHO that doesn't improve the readability.

Reply:
Ah yes - you could reduce the refactoring needed by just defining the record as normal (that is...

```
type negotiated_kex_state = { client_version : string ; client_kex : Ssh.kexinit ;
                        server_version : string ; server_kex : Ssh.kexinit ;
                        negotiated : Kex.negotiation ;
                        dh_secret : Mirage_crypto_pk.Dh.secret ;
                        dh_share : Ssh.mpint }
type kex_state =
| Negotiated_kex of negotiated_kex_state
```

This does prevent any possible mixup of the same `string` values in the `handle_kexdh_reply`, which is what was worrying me in the current diff.  Alternatively, handle_kexdh_reply could use labelled arguments to distinguish the arguments that have the same type.